File: src/main/java/org/dedda/bratwurst/parse/ConditionParser.java
package org.dedda.bratwurst.parse;

import org.dedda.bratwurst.lang.BWExpression;
import org.dedda.bratwurst.lang.BWInstruction;
import org.dedda.bratwurst.lang.Condition;

import java.util.LinkedList;
import java.util.List;

import static org.dedda.bratwurst.parse.Patterns.CONDITION_END;
import static org.dedda.bratwurst.parse.Patterns.CONDITION_SEPARATOR;

/**
 * Created by dedda on 10/19/15.
 *
 * @author dedda
 */
public class ConditionParser {

    public Condition parse(final String[] lines, final int begin) {
        final int end = getEnd(lines, begin);
        final int separator = getSeparator(lines, begin);
        final String conditionLine = lines[begin];
        final List<BWInstruction> trueInstructions = new LinkedList<>();
        final List<BWInstruction> falseInstructions = new LinkedList<>();
        for (int i = begin+1; i < separator; i++) {
            trueInstructions.add(new InstructionParser().parse(lines[i], i));
        }
        for (int i = separator+1; i < end; i++) {
            falseInstructions.add(new InstructionParser().parse(lines[i], i));
        }
        final BWExpression condition = new ExpressionParser().parse(conditionLine.substring(2, conditionLine.length() - 3), begin);
        final BWInstruction[] trueInstructionsArray = new BWInstruction[trueInstructions.size()];
        trueInstructions.toArray(trueInstructionsArray);
        final BWInstruction[] falseInstructionsArray = new BWInstruction[falseInstructions.size()];
        falseInstructions.toArray(falseInstructionsArray);
        return new Condition(begin, condition, trueInstructionsArray, falseInstructionsArray);
    }

    public int getEnd(final String[] lines, final int begin) {
        for (int i = begin; i < lines.length; i++) {
            if (lines[i].matches(CONDITION_END)) {
                return i;
            }
        }
        throw new RuntimeException("No condition end found!");
    }

    private int getSeparator(final String[] lines, final int begin) {
        for (int i = begin; i < lines.length; i++) {
            if (lines[i].matches(CONDITION_SEPARATOR)) {
                return i;
            }
        }
        throw new RuntimeException("No condition separator found!");
    }

}

File: src/main/java/org/dedda/bratwurst/parse/ConditionParser.kt
package org.dedda.bratwurst.parse

import org.dedda.bratwurst.lang.Condition
import org.dedda.bratwurst.parse.Patterns.CONDITION_END
import org.dedda.bratwurst.parse.Patterns.CONDITION_SEPARATOR
import java.util.stream.IntStream

/**
 * Created by dedda on 10/19/15.
 *
 * @author dedda
 */
class ConditionParser {
    fun parse(lines: List<String>, begin: Int): Condition {
        val end = findEnd(lines, begin)
        val separator = findSeparator(lines, begin)
        val conditionLine = lines[begin]
        val trueInstructions = (begin + 1 until separator)
            .mapNotNull { InstructionParser().parse(lines[it], it) }
            .toList()
        val falseInstructions = (separator + 1 until end)
            .mapNotNull { InstructionParser().parse(lines[it], it) }
            .toList()
        val condition = ExpressionParser().parse(conditionLine.substring(2, conditionLine.length - 3), begin)
        return Condition(begin, condition, trueInstructions, falseInstructions)
    }

    fun findEnd(lines: List<String>, begin: Int): Int {
        val pattern = Regex(CONDITION_END)
        return IntStream.range(begin, lines.size)
            .filter { line: Int -> lines[line].matches(pattern) }
            .findFirst()
            .orElseThrow { RuntimeException("No condition end found!") }
    }

    private fun findSeparator(lines: List<String>, begin: Int): Int {
        val pattern = Regex(CONDITION_SEPARATOR)
        return IntStream.range(begin, lines.size)
            .filter { line: Int -> lines[line].matches(pattern) }
            .findFirst()
            .orElseThrow { RuntimeException("No condition separator found!") }
    }
}